File: sdk/telemetry/python/examples/fetch.py
#!/usr/bin/env python3
"""Example CLI that fetches and displays telemetry data."""

import argparse
import sys
import time

from solders.pubkey import Pubkey  # type: ignore[import-untyped]

from serviceability.client import Client as ServiceabilityClient
from telemetry.client import Client as TelemetryClient


def main() -> None:
    parser = argparse.ArgumentParser(description="Fetch telemetry data")
    parser.add_argument(
        "--env",
        default="mainnet-beta",
        choices=["mainnet-beta", "testnet", "devnet", "localnet"],
        help="Environment to connect to",
    )
    parser.add_argument(
        "--epoch",
        type=int,
        default=0,
        help="Epoch to fetch samples for (0 = try recent epochs)",
    )
    args = parser.parse_args()

    print(f"Fetching telemetry data from {args.env}...\n")

    # First, get serviceability data to discover devices and links
    svc_client = ServiceabilityClient.from_env(args.env)
    svc_data = svc_client.get_program_data()

    print("=== Network Overview ===")
    print(f"Devices: {len(svc_data.devices)}")
    print(f"Links:   {len(svc_data.links)}")
    print()

    if not svc_data.links:
        print("No links found - no telemetry data to fetch.")
        return

    # Build device code map for display
    device_codes: dict[str, str] = {}
    device_pks: dict[str, Pubkey] = {}
    for dev in svc_data.devices:
        pk_str = str(dev.owner)  # Using owner as pubkey proxy
        device_codes[pk_str] = dev.code
        device_pks[dev.code] = dev.owner

    # Create telemetry client
    tel_client = TelemetryClient.from_env(args.env)

    # Determine which epoch to try
    target_epoch = args.epoch
    if target_epoch == 0:
        # Rough approximation based on current time
        target_epoch = int(time.time() / 432000)

    print(f"=== Device Latency Samples (epoch {target_epoch}) ===")

    samples_found = 0
    for link in svc_data.links:
        side_a_pk = link.side_a_pub_key
        side_z_pk = link.side_z_pub_key
        link_pk = link.owner  # Using owner as link pubkey proxy

        # Find device codes
        side_a_code = "unknown"
        side_z_code = "unknown"
        for dev in svc_data.devices:
            if str(dev.owner) == str(side_a_pk):
                side_a_code = dev.code
            if str(dev.owner) == str(side_z_pk):
                side_z_code = dev.code

        # Try both directions
        for origin_pk, target_pk, o_code, t_code in [
            (side_a_pk, side_z_pk, side_a_code, side_z_code),
            (side_z_pk, side_a_pk, side_z_code, side_a_code),
        ]:
            try:
                samples = tel_client.get_device_latency_samples(
                    origin_pk, target_pk, link_pk, target_epoch
                )
            except Exception:
                # Account likely doesn't exist for this epoch
                continue

            samples_found += 1
            sample_count = len(samples.samples)

            if sample_count == 0:
                print(f"  {o_code} -> {t_code} ({link.code}): initialized, no samples yet")
                continue

            # Calculate stats
            total = sum(samples.samples)
            min_val = min(samples.samples)
            max_val = max(samples.samples)
            avg_us = total / sample_count
            avg_ms = avg_us / 1000.0
            min_ms = min_val / 1000.0
            max_ms = max_val / 1000.0

            print(
                f"  {o_code} -> {t_code} ({link.code}): {sample_count} samples, "
                f"avg {avg_ms:.2f}ms, min {min_ms:.2f}ms, max {max_ms:.2f}ms"
            )

    if samples_found == 0:
        print(f"  No samples found for epoch {target_epoch}. Try a different epoch with --epoch flag.")

    print()
    print("Done.")


if __name__ == "__main__":
    main()

File: sdk/telemetry/typescript/examples/fetch.ts
#!/usr/bin/env bun
/**
 * Example CLI that fetches and displays telemetry data.
 *
 * For a full example with device discovery, use the serviceability SDK
 * to fetch devices and links, then use their public keys here.
 *
 * Usage:
 *   bun run examples/fetch.ts --env testnet --origin <pubkey> --target <pubkey> --link <pubkey> --epoch 12345
 *
 * Or run the serviceability example first to discover devices and links.
 */

import { PublicKey } from "@solana/web3.js";
import { Client } from "../telemetry/client.js";

async function main() {
  const args = process.argv.slice(2);
  let env = "mainnet-beta";
  let originPK: string | null = null;
  let targetPK: string | null = null;
  let linkPK: string | null = null;
  let epoch: number | null = null;

  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case "--env":
        env = args[++i];
        break;
      case "--origin":
        originPK = args[++i];
        break;
      case "--target":
        targetPK = args[++i];
        break;
      case "--link":
        linkPK = args[++i];
        break;
      case "--epoch":
        epoch = parseInt(args[++i], 10);
        break;
    }
  }

  const validEnvs = ["mainnet-beta", "testnet", "devnet", "localnet"];
  if (!validEnvs.includes(env)) {
    console.error(`Invalid environment: ${env}`);
    process.exit(1);
  }

  if (!originPK || !targetPK || !linkPK || epoch === null) {
    console.log("Usage: fetch.ts --env <env> --origin <pubkey> --target <pubkey> --link <pubkey> --epoch <num>");
    console.log();
    console.log("This example requires specific device and link public keys.");
    console.log("Run the serviceability SDK example first to discover available devices and links.");
    console.log();
    console.log("Example (with placeholder values - replace with real pubkeys):");
    console.log("  bun run examples/fetch.ts --env testnet \\");
    console.log("    --origin 11111111111111111111111111111111 \\");
    console.log("    --target 22222222222222222222222222222222 \\");
    console.log("    --link 33333333333333333333333333333333 \\");
    console.log("    --epoch 12345");
    process.exit(0);
  }

  console.log(`Fetching telemetry data from ${env}...\n`);

  const client = Client.forEnv(env);

  try {
    const samples = await client.getDeviceLatencySamples(
      new PublicKey(originPK),
      new PublicKey(targetPK),
      new PublicKey(linkPK),
      epoch,
    );

    console.log("=== Device Latency Samples ===");
    console.log(`Epoch: ${samples.epoch}`);
    console.log(`Origin Device: ${new PublicKey(samples.originDevicePK).toBase58()}`);
    console.log(`Target Device: ${new PublicKey(samples.targetDevicePK).toBase58()}`);
    console.log(`Link: ${new PublicKey(samples.linkPK).toBase58()}`);
    console.log(`Sampling Interval: ${samples.samplingIntervalMicroseconds}us`);
    console.log(`Sample Count: ${samples.samples.length}`);

    if (samples.samples.length > 0) {
      const total = samples.samples.reduce((a, b) => a + b, 0);
      const min = Math.min(...samples.samples);
      const max = Math.max(...samples.samples);
      const avg = total / samples.samples.length;

      console.log();
      console.log("Statistics:");
      console.log(`  Average: ${(avg / 1000).toFixed(2)}ms`);
      console.log(`  Min: ${(min / 1000).toFixed(2)}ms`);
      console.log(`  Max: ${(max / 1000).toFixed(2)}ms`);
    }

  } catch (err) {
    if ((err as Error).message === "Account not found") {
      console.log("No samples found for the specified parameters.");
      console.log("The account may not exist for this epoch or device pair.");
    } else {
      throw err;
    }
  }

  console.log();
  console.log("Done.");
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
